File: libs/OrbMaths.c
#include "OrbMaths.h"

// ------------------------------------------------------------------

// double cbrt(double x) {
// 	return pow(x, 1.0/3.0);
// }

double OrbitalPeriod(double MeanMotion) {
	double Period;

	Period = (double)(EARTH_DAY_LENGTH / MeanMotion);

	return Period;
}

uint64_t SemiMajorAxis(double Period) {
	double SMA;

	SMA = cbrt(pow((Period / (2.0 * M_PI)), 2.0) * EARTH_MU);

	//SMA = cbrt((G * EARTH_MASS * pow(Period, 2))/(4.0*pow(M_PI, 2.0)));

	if ((SMA >= 0) & (SMA <= 0xffffffffffffffffULL)) {
		return (uint64_t)SMA;
	} else {
		return 0;
	}
}



uint64_t Apoapsis(double Eccentricity, uint64_t SemiMajorAxis) {
	uint64_t Apoapsis = 0;

	Apoapsis = (uint64_t)(SemiMajorAxis * (1.0 + Eccentricity));

	return Apoapsis;
}

uint64_t Periapsis(double Eccentricity, uint64_t SemiMajorAxis) {
	uint64_t Periapsis = 0;

	Periapsis = (uint64_t)(SemiMajorAxis * (1.0 - Eccentricity));

	return Periapsis;
}

uint64_t OrbAlt(double Eccentricity, uint64_t SemiMajorAxis, double EccentricAnomaly) {
	return (uint64_t)(SemiMajorAxis * (1.0 - Eccentricity * cos(EccentricAnomaly)));
}

uint64_t OrbAltTA(double Eccentricity, uint64_t SemiMajorAxis, double TrueAnomaly) {
	return (uint64_t)(SemiMajorAxis * (1.0 - pow(Eccentricity, 2.0)) / (1.0 + Eccentricity * cos(TrueAnomaly)));
}



double KeplerEquation(double EccentricAnomaly, double Eccentricity) {
	return EccentricAnomaly - Eccentricity * sin(EccentricAnomaly);
}

double KeplerPrime(double EccentricAnomaly, double Eccentricity) {
	return 1.0 - Eccentricity * cos(EccentricAnomaly);
}

double AngularSpeed(double SemiMajorAxis) {
	return sqrt(EARTH_MU/pow(SemiMajorAxis, 3));
}

double TrueAnomaly(double Eccentricity, double EccentricAnomaly) {
	return 2.0 * atan(sqrt((1.0f+Eccentricity)/(1.0f-Eccentricity) * tan(EccentricAnomaly/2.0)));
}



double OrbSpeed(uint64_t altitude, uint64_t SemiMajorAxis) {
	double speed;

	speed = sqrt(EARTH_MU * (2/((double)altitude) - 1/((double)SemiMajorAxis)));

	return speed;
}

double Lorentz(double v) {
	const double c_sqr = pow(299792458.0, 2.0);

	return 1.0 / sqrt(1 - pow(v, 2.0)/c_sqr);
}

File: PythonTLE/OrbMaths.py
from math import *

EARTH_DAY_LENGTH: float = 86400

G: float = 6.67428e-11#428e-11
EARTH_MASS: float = 5.9722e24
POLAR_RADIUS: int = 6356752
EQUATORIAL_RADIUS: int = 6378137

EARTH_RADIUS: float = 6371000#(2*EQUATORIAL_RADIUS + POLAR_RADIUS)//3

EARTH_MU: int = int(G*EARTH_MASS)

C: int = 299_792_458

def OrbitalPeriod(MeanMotion: float) -> float:
	Period: float = 0.0

	Period = EARTH_DAY_LENGTH/MeanMotion

	return Period

def SemiMajorAxis(Period: float) -> int:
	SMA: int = 0.0

	SMA = int(cbrt((Period/(2*pi))**2 * EARTH_MU))

	return SMA



def Apogee(Eccentricity: float, SMA: int) -> int:
	Apogee: int = 0

	Apogee = int(SMA * (1 + Eccentricity))

	return Apogee

def Perigee(Eccentricity: float, SMA: int) -> int:
	Perigee: int = 0

	Perigee = int(SMA * (1 - Eccentricity))

	return Perigee

def OrbAlt(Eccentricity: float, SMA: int, E: float) -> int:
	OrbAlt: int = 0

	OrbAlt = int(SMA * (1 - Eccentricity * cos(E)))

	return OrbAlt

def OrbAltTA(Eccentricity: float, SMA: int, TrueAnomaly: float) -> float:
	return SMA * (1-Eccentricity**2) / (1+Eccentricity * cos(TrueAnomaly))


def KeplerEquation(E: float, e: float) -> float:
	return E - e*sin(E)

def KeplerPrime(E: float, e: float) -> float:
	return 1 - e*cos(E)

def AngularSpeed(SMA: int) -> float:
	return sqrt(EARTH_MU/SMA**3)

def Beta(Eccentricity: float) -> float:
	return Eccentricity/(1+sqrt(1-Eccentricity**2))

# def TrueAnomaly(Eccentricity: float, EccentricAnomaly: float) -> float:
# 	return 2 * atan(sqrt((1+Eccentricity)/(1-Eccentricity) * tan(EccentricAnomaly/2)))

def TrueAnomaly(Eccentricity: float, EccentricAnomaly: float) -> float:
	return EccentricAnomaly + 2 * atan((Beta(Eccentricity) * sin(EccentricAnomaly))/(1 - Beta(Eccentricity) * cos(EccentricAnomaly)))



def OrbSpeed(altitude: int, SMA: int) -> float:
	speed: float = 0.0

	speed = sqrt(EARTH_MU * (2/altitude - 1/SMA))

	return speed

def Lorentz(speed: float) -> float:
	lorentz: float = 0.0

	lorentz = 1 / sqrt(1 - speed**2/C**2)

	return lorentz
